### PR TITLE
TFLite: Div op Neon optimization

### PR DESCRIPTION
**Pull request by mwtarnowski**
Added float32 division optimized with Neon SIMD instructions.

### CODE REVIEW

**Comment by mwtarnowski** [comment]
@gbaned, do you know when this PR will be processed? Are there any problems with it?

**Comment by gbaned** [comment]
@mwtarnowski Thanks for your contribution, waiting for a googler to approve the changes.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @aselle, @jianlijianli: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 15 days with no activity and the `awaiting review` label has been applied.

**Comment by mwtarnowski** [comment]
@aselle, @jianlijianli, do you know (more or less) when this code could be reviewed?

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Comment by jdduke** [comment]
Can you provide some additional details on the associated speedup? and how you benchmarked/verified the speedup? Thanks!

**Comment by gbaned** [comment]
@mwtarnowski Can you please resolve conflicts? Thanks!

**Comment by mwtarnowski** [comment]
@gbaned Done.

@jdduke Below are results of running [benchmark tool](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/tools/benchmark) on Google Pixel 3 XL

`./benchmark_model --graph=div.tflite --num_threads=1`

for a simple model `div.tflite` consisting only of elementwise division of two tensors of shape (1,512,512,128).
visualize_div

current implementation:
```
Initialized session in 0.497ms
Running benchmark for at least 1 iterations and at least 0.5 seconds but terminate if exceeding 150 seconds.
count=3 first=155086 curr=121309 min=121309 max=155086 avg=132658 std=15859

Running benchmark for at least 50 iterations and at least 1 seconds but terminate if exceeding 150 seconds.
count=50 first=120700 curr=121057 min=120227 max=122100 avg=120982 std=415

Memory usage: max resident set size = 640.383 MB
Memory usage: total malloc-ed memory = 704.17 MB
Average inference timings in us: Warmup: 132658, Init: 497, no stats: 120982
```
proposed optimization:
```
Initialized session in 0.447ms
Running benchmark for at least 1 iterations and at least 0.5 seconds but terminate if exceeding 150 seconds.
count=5 first=89585 curr=40311 min=40149 max=89585 avg=50116.8 std=19734

Running benchmark for at least 50 iterations and at least 1 seconds but terminate if exceeding 150 seconds.
count=50 first=40277 curr=41256 min=40021 max=41256 avg=40326 std=181

Memory usage: max resident set size = 640.387 MB
Memory usage: total malloc-ed memory = 704.17 MB
Average inference timings in us: Warmup: 50116.8, Init: 447, no stats: 40326
```
Please let me know if you need to see results for other devices.

**Comment by gbaned** [comment]
@mwtarnowski Still, conflicts appearing. Could you please resolve those? Thanks!

**Comment by mwtarnowski** [comment]
@gbaned Resolved.

File: tensorflow/lite/kernels/internal/optimized/optimized_ops.h [hunk 1]
```diff
@@ -2629,6 +2629,84 @@ inline void BroadcastMulDispatch(
...
+                const RuntimeShape& input2_shape, const float* input2_data,
+                const RuntimeShape& output_shape, float* output_data) {
+  gemmlowp::ScopedProfilingLabel label("Div");
+  const float output_activation_min = params.float_activation_min;
+  const float output_activation_max = params.float_activation_max;
+
+  int i = 0;
+  const int size = MatchingFlatSize(input1_shape, input2_shape, output_shape);
+#ifdef USE_NEON
+  static constexpr int kNewtonSteps = 2;
```
Comment (jianlijianli) [inline tensorflow/lite/kernels/internal/optimized/optimized_ops.h]: Could you please add a comment on this parameter? How is it used and what is the accuracy implication if kNewtonStep is 2?

**Review by jianlijianli — DISMISSED** [review]
Thanks for the contribution. Added a comment on the kNewtonSteps parameter.

**Comment by gbaned** [comment]
@mwtarnowski Could you please check reviewer comments and keep us posted. Thanks!

File: tensorflow/lite/kernels/internal/optimized/optimized_ops.h (hunk 1 above)
Comment (mwtarnowski) [reply tensorflow/lite/kernels/internal/optimized/optimized_ops.h]: Added a comment.

`kNewtonSteps` stands for a number of Newton-Raphson iterations used to refine the initial estimate of the reciprocal given by `vrecpeq_f32` instruction.

The list below shows implication of this parameter on maximum and average relative errors between the results of NEON multiplication by the reciprocal and the actual result of floating point division, where both of operands are random numbers in the range (2^-16, 2^16).

- kNewtonStep=0: max=2.76e-03 avg=9.70e-04
- kNewtonStep=1: max=7.57e-06 avg=1.39e-06
- kNewtonStep=2: max=1.52e-07 avg=3.49e-08
- kNewtonStep=3: max=1.19e-07 avg=2.52e-08

**Comment by mwtarnowski** [comment]
@gbaned Updated.

**Review by jianlijianli — APPROVED** [review]
Thanks Michal.

**Pull request merged**